src/algorithm/distance.rs

```diff
@@ -27,13 +72,261 @@ impl<T> Distance<T, Point<T>> for Point<T>
     }
 }
 
+// Return minimum distance between a Point and a Line segment
+// This is a helper for Point-to-LineString and Point-to-Polygon distance
+// adapted from http://stackoverflow.com/a/1501725/416626
+fn line_segment_distance<T>(point: &Point<T>, start: &Point<T>, end: &Point<T>) -> T
+    where T: Float + ToPrimitive
+{
+    let dist_squared = pow(start.distance(end), 2);
+    // Implies that start == end
+    if dist_squared == T::zero() {
+        return pow(point.distance(start), 2);
+    }
+    // Consider the line extending the segment, parameterized as start + t (end - start)
+    // We find the projection of the point onto the line
+    // This falls where t = [(point - start) . (end - start)] / |end - start|^2, where . is the dot product
+    // We clamp t from [0.0, 1.0] to handle points outside the segment start, end
+    let t = T::zero().max(T::one().min((*point - *start).dot(&(*end - *start)) / dist_squared));
+    let projected = Point::new(start.x() + t * (end.x() - start.x()),
+                               start.y() + t * (end.y() - start.y()));
+    point.distance(&projected)
+}
+
+#[derive(PartialEq, Debug)]
```

Comment:
`Ord`, `PartialOrd`, and `Eq` need custom `impl`s (lines 111, 118, 125) for these, in order for the `BinaryHeap` to work as min-heap, so deriving them causes a conflict…